packages/eslint-plugin/lib/rules/restrict-comparison-operands.js

```diff
@@ -0,0 +1,161 @@
+/* eslint-env node */
+/* eslint no-bitwise: ["off"] */
+
+'use strict';
+
+const ts = require('typescript');
+const tsutils = require('tsutils');
+const { ESLintUtils } = require('@typescript-eslint/utils');
+
+const COMPARABLE_TYPES = ['number', 'bigint', 'string', 'any'];
+const NONCOMPARABLE = Symbol('non-comparable type');
+
+const getTypeFlags = type => {
+  let flags = 0;
+  for (const subType of tsutils.unionTypeParts(type)) {
+    flags |= subType.flags;
+  }
+  return flags;
+};
+
+const createRule = ESLintUtils.RuleCreator(
+  name =>
+    `https://github.com/endojs/endo/blob/master/packages/eslint-plugin/lib/rules/${name}.js`,
+);
+
+module.exports = createRule({
+  name: 'restrict-comparison-operands',
+  meta: {
+    docs: {
+      description:
+        'require both operands of a comparison operator (`<`, `>`, `<=`, `>=`) to be compatible types, either both primitive strings or both primitive numerics (number or bigint)',
+      category: 'Possible Errors',
+      recommended: true,
+      requiresTypeChecking: true,
+    },
+    type: 'problem',
+    fixable: null,
+    messages: {
+      mismatch: 'Comparison of mismatched types',
+      invalidType: 'Comparison of invalid type(s)',
+      unknownType: 'Comparison against unknown type',
+    },
+    schema: [
+      {
+        type: 'object',
+        additionalProperties: false,
+        properties: {
+          allowUnknown: { type: 'boolean' },
+          // TODO: `allowMixedNumerics: { type: "boolean" }`?
+        },
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      allowUnknown: false,
+    },
+  ],
+  create(context, [{ allowUnknown }]) {
+    // Loosely follows
+    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/restrict-plus-operands.ts
+
+    const { parserServices } = context;
+    const typeChecker = parserServices?.program.getTypeChecker();
+    if (
+      !parserServices ||
+      !parserServices.hasFullTypeInformation ||
+      !parserServices.esTreeNodeToTSNodeMap ||
+      !typeChecker
+    ) {
+      // Intentionally avoid a SES `assert` dependency.
+      throw new Error(
+        `Missing full type information. Make sure eslint configuration "parser" uses @typescript-eslint/parser parser and "parserOptions" specifies a "project" TypeScript configuration that includes each file subject to this rule.`,
+      );
+    }
+
+    const comparableTypeOf = type => {
+      if (type.isNumberLiteral()) {
+        return 'number';
+      } else if (
+        type.isStringLiteral() ||
+        getTypeFlags(type) & ts.TypeFlags.TemplateLiteral
```

Comment:
Does this exclude tagged template literals?

Reply:
It does, but your comment prompted me to dig deeper and ultimately ignore literal vs. non-literal in favor of TypeFlags.{Number,String,BigInt}Like bitmasks (in which e.g. `'foo'`, ``` `bar${suffix}` ```, and `'baz' + String(getSuffix())` are all equally well covered).